File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/Collections.kt
package sudokusolver.kotlin

import java.util.EnumSet

inline fun <reified T : Enum<T>> enumIntersect(vararg sets: EnumSet<T>): EnumSet<T> =
    when (sets.size) {
        0 -> EnumSet.noneOf(T::class.java)
        1 -> sets.single()

        else -> EnumSet.allOf(T::class.java).also { intersection ->
            sets.forEach { set -> intersection.retainAll(set) }
        }
    }

inline fun <reified T : Enum<T>> enumUnion(vararg sets: EnumSet<T>): EnumSet<T> =
    when (sets.size) {
        0 -> EnumSet.noneOf(T::class.java)
        1 -> sets.single()
        else -> EnumSet.noneOf(T::class.java).also { union -> sets.forEach { set -> union += set } }
    }

infix fun <T : Enum<T>> EnumSet<T>.enumMinus(other: EnumSet<T>): EnumSet<T> =
    EnumSet.copyOf(this).apply { this -= other }

inline fun <reified T : Enum<T>, reified R : Enum<R>> Iterable<Pair<T, R>>.enumUnzip(): Pair<EnumSet<T>, EnumSet<R>> =
    fold(EnumSet.noneOf(T::class.java) to EnumSet.noneOf(R::class.java)) { sets, pair ->
        sets.first += pair.first
        sets.second += pair.second
        sets
    }

fun <T> List<T>.zipEveryPair(): List<Pair<T, T>> =
    mapIndexed { firstIndex, first -> drop(firstIndex + 1).map { second -> first to second } }.flatten()

fun <T> List<T>.zipEveryTriple(): List<Triple<T, T, T>> =
    mapIndexed { firstIndex, first ->
        withIndex().drop(firstIndex + 1).flatMap { (secondIndex, second) ->
            drop(secondIndex + 1).map { third -> Triple(first, second, third) }
        }
    }.flatten()

data class Quad<out A, out B, out C, out D>(val first: A, val second: B, val third: C, val fourth: D)

fun <T> List<T>.zipEveryQuad(): List<Quad<T, T, T, T>> =
    mapIndexed { firstIndex, first ->
        withIndex().drop(firstIndex + 1).flatMap { (secondIndex, second) ->
            withIndex().drop(secondIndex + 1).flatMap { (thirdIndex, third) ->
                drop(thirdIndex + 1).map { fourth -> Quad(first, second, third, fourth) }
            }
        }
    }.flatten()

File: sudokusolver_scala/src/main/scala/sudokusolver/scala/Collections.scala
package sudokusolver.scala

extension [T](seq: IndexedSeq[T])
  def zipEveryPair: IndexedSeq[(T, T)] =
    for
      (first, firstIndex) <- seq.zipWithIndex
      second <- seq.drop(firstIndex + 1)
    yield first -> second

  def zipEveryTriple: IndexedSeq[(T, T, T)] =
    for
      (first, firstIndex) <- seq.zipWithIndex
      (second, secondIndex) <- seq.zipWithIndex.drop(firstIndex + 1)
      third <- seq.drop(secondIndex + 1)
    yield (first, second, third)

  def zipEveryQuad: IndexedSeq[(T, T, T, T)] =
    for
      (first, firstIndex) <- seq.zipWithIndex
      (second, secondIndex) <- seq.zipWithIndex.drop(firstIndex + 1)
      (third, thirdIndex) <- seq.zipWithIndex.drop(secondIndex + 1)
      fourth <- seq.drop(thirdIndex + 1)
    yield (first, second, third, fourth)